### PR TITLE
Add tests for create2 in creator

### PR DESCRIPTION
**Pull request by sorpaas**
rel https://github.com/paritytech/parity-ethereum/pull/9277

* Fixes all scripts to use `/usr/bin/env`, otherwise it breaks on non-FHS Linux.
* Fixes `build-all.sh` to build for newest version of `wasm-build`.
* Allow custom features (like `kip4`) to be enabled for specific tests, in `gen`.
* Add tests for create2 in `creator.rs`.

### CODE REVIEW

File: src/alloc.rs [hunk 1]
```diff
@@ -12,4 +12,7 @@ pub fn call() {
 		data.extend_from_slice(&[5u8; 400*1024][..]);
 		data
 	});
-}
\ No newline at end of file
+}
+
+#[no_mangle]
+pub fn deploy() { }
```
Comment (pepyakin) [inline src/alloc.rs]: I wonder why these are needed?

**Review by pepyakin — APPROVED** [review]
Nice!

File: src/alloc.rs (hunk 1 above)
Comment (sorpaas) [reply src/alloc.rs]: `wasm-build` would complain if no `deploy` symbol is found: https://github.com/paritytech/wasm-utils/blob/master/src/build.rs#L110
In the binary, we always set [constructor to true](https://github.com/paritytech/wasm-utils/blob/master/src/build.rs#L108) and build the deploy wrapped version.

**Comment by sorpaas** [comment]
Only merge after https://github.com/paritytech/pwasm-ethereum/pull/12 -- we would need to change the `pwasm-ethereum` branch from `kip4` to `master`.

File: gen/main.rs [hunk 2]
```diff
@@ -3,23 +3,26 @@ use std::path::PathBuf;
...
 [package]
 name = "wasm-tests"
 version = "0.1.0"
 authors = ["NikVolf <nikvolf@gmail.com>"]
 
 [dependencies]
-pwasm-std = "0.9.0"
-pwasm-ethereum = "0.5.0"
-bigint = { version = "4", default-features = false }
+pwasm-std = {}
```
Comment (NikVolf) [inline gen/main.rs]: could just be updated to 0.10 for all?

File: gen/main.rs [hunk 3]
```diff
@@ -3,23 +3,26 @@ use std::path::PathBuf;
...
+pwasm-ethereum = {}
```
Comment (NikVolf) [inline gen/main.rs]: could be just 0.6 ?

File: src/creator.rs [hunk 4]
```diff
@@ -2,16 +2,33 @@
...
-		ret(&addr[..]);
+	let mut r = [0u8; 40];
+	if let Ok(addr) = create(value() / U256::from(2), &input()) {
+		logger::debug("Created contract with code");
+		(&mut r[0..20]).copy_from_slice(&addr[..]);
 	} else {
 		logger::debug("Error creating contract");
 	}
+
+	if let Ok(addr) = create2(value() / U256::from(2), H256::default(), &input()) {
```
Comment (NikVolf) [inline src/creator.rs]: maybe let's use some non-zero salt to actually test the argument reaches the runtime after?

File: src/dispersion.rs [hunk 5]
```diff
@@ -23,4 +23,7 @@ pub fn call() {
 	}
 
 	ret(&dispersed[..])
-}
\ No newline at end of file
+}
+
+#[no_mangle]
+pub fn deploy() { }
```
Comment (NikVolf) [inline src/dispersion.rs]: there should not be `deploy` in any of those tests

with deploy, they will get packed as constructors, while they need to get packed as raw wasm module

File: src/alloc.rs (hunk 1 above)
Comment (NikVolf) [reply src/alloc.rs]: @sorpaas 

there is a bug in wasm-build 0.3.0 with this constructor=true always

you might try pwasm-utils-cli 0.2.0 (`cargo install pwasm-utils-cli --version 0.2.0`) or wait until @fckt fix it

Comment (lexfrl) [reply src/alloc.rs]: https://github.com/paritytech/wasm-utils/pull/100

File: src/dispersion.rs (hunk 5 above)
Comment (sorpaas) [reply src/dispersion.rs]: @NikVolf With no deploy `wasm-build` would refuse to pack those modules: https://github.com/paritytech/wasm-utils/blob/master/src/build.rs#L110

Comment (NikVolf) [reply src/dispersion.rs]: should be fixed now

Comment (NikVolf) [reply src/dispersion.rs]: (reinstall pwasm-utils-cli)

Comment (sorpaas) [reply src/dispersion.rs]: Removed all `deploy` symbols!

**Comment by NikVolf** [comment]
You might want to avoid including all recompiled wasm binaries in the diff and include only `creator.wasm`

Otherwise you will have to update gas costs for all tests in `parity-ethereum` repo

but it's up to you!

**Comment by lexfrl** [comment]
> Otherwise you will have to update gas costs for all tests in parity-ethereum repo

And in case you'll decide to do so, please update the `rust-toolchain` file to use `nightly-2018-07-24` to sync in with other wasm repos, like https://github.com/paritytech/pwasm-std/blob/master/rust-toolchain

**Comment by sorpaas** [comment]
Some of the gas costs are already changed due to update to `pwasm-ethereum`. So I think I would just fix the gas costs. And updated `rust-toolchain`!

**Pull request merged**